### PR TITLE
Update to ClimaComms v0.5.0

### PR DESCRIPTION
**Pull request by sriharshakandala**
Update src

- [x] Code follows the [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) OR N/A.
- [x] Unit tests are included OR N/A.
- [x] Code is exercised in an integration test OR N/A.
- [x] Documentation has been added/updated OR N/A.

### CODE REVIEW

**Comment by sriharshakandala** [comment]
bors try

**Comment by sriharshakandala** [comment]
bors try

**Comment by sriharshakandala** [comment]
bors try

**Comment by sriharshakandala** [comment]
bors try

File: src/Spaces/pointspace.jl [hunk 1]
```diff
@@ -16,8 +16,10 @@ end
 ClimaComms.device(space::PointSpace) = ClimaComms.device(space.context)
 ClimaComms.context(space::PointSpace) = space.context
 
-PointSpace(x::Geometry.LocalGeometry) = PointSpace(ClimaComms.CPUDevice(), x)
-PointSpace(x::Geometry.AbstractPoint) = PointSpace(ClimaComms.CPUDevice(), x)
+PointSpace(x::Geometry.LocalGeometry) =
+    PointSpace(ClimaComms.AbstractCPUDevice(), x)
```
Comment (charleskawczynski) [inline src/Spaces/pointspace.jl:20]:
```suggestion
    PointSpace(ClimaComms.CPUSingleThreaded(), x)
```

File: src/Spaces/pointspace.jl [hunk 2]
```diff
@@ -16,8 +16,10 @@ end
...
+PointSpace(x::Geometry.AbstractPoint) =
+    PointSpace(ClimaComms.AbstractCPUDevice(), x)
```
Comment (charleskawczynski) [inline src/Spaces/pointspace.jl:22]:
```suggestion
    PointSpace(ClimaComms.CPUSingleThreaded(), x)
```

**Comment by sriharshakandala** [comment]
bors try

**Comment by sriharshakandala** [comment]
bors try

**Comment by sriharshakandala** [comment]
bors try

**Comment by simonbyrne** [comment]
It would be good to make the threading changes as well: e.g. get rid of `enable_threading()` and use the device to determine whether or not to use threading.

**Comment by simonbyrne** [comment]
e.g. 

```julia
function byslab(fn, space::Spaces.AbstractSpace)
    byslab(fn, ClimaComms.device(space), space)
end

function byslab(fn, ::ClimaComms.CPUSingleThreaded, space::Spaces.AbstractSpectralElementSpace)
    Nh = Topologies.nlocalelems(space.topology)::Int
    @inbounds for h in 1:Nh
        fn(SlabIndex(nothing, h))
    end
end
function byslab(fn, ::ClimaComms.CPUMultiThreaded, space::Spaces.AbstractSpectralElementSpace)
    Nh = Topologies.nlocalelems(space.topology)::Int
    @inbounds Threads.@threads for h in 1:Nh
        fn(SlabIndex(nothing, h))
    end
end
```

**Comment by sriharshakandala** [comment]
bors try

**Comment by charleskawczynski** [comment]
> `@inbounds Threads.@threads for h in 1:Nh`

There's a slack discussion regarding `@inbounds Threads.@threads`: https://julialang.slack.com/archives/C67TK21LJ/p1670593631969889

**Comment by charleskawczynski** [comment]
We should be able to eliminate `enable_threading()` in this PR, too.

**Comment by charleskawczynski** [comment]
How should we adjust

```julia
function Base.copyto!(
    dest::VIFH{S, Ni},
    bc::Base.Broadcast.Broadcasted{VIFHStyle{Ni, A}},
) where {S, Ni, A}
    if enable_threading()
        return _threaded_copyto!(dest, bc)
    end
    return _serial_copyto!(dest, bc)
end
```
?

Do we just go through and adjust all the callers?

**Comment by sriharshakandala** [comment]
> How should we adjust
> 
> ```julia
> function Base.copyto!(
>     dest::VIFH{S, Ni},
>     bc::Base.Broadcast.Broadcasted{VIFHStyle{Ni, A}},
> ) where {S, Ni, A}
>     if enable_threading()
>         return _threaded_copyto!(dest, bc)
>     end
>     return _serial_copyto!(dest, bc)
> end
> ```
> 
> ?
> 
> Do we just go through and adjust all the callers?

One option is to check
```
if array_type(dest) isa Array && Threads.nthreads() > 1
     do multhreaded_copy
else
    do singlethreaded_copy
end

**Comment by charleskawczynski** [comment]
@sriharshakandala suggested ditching the two threaded functions for now, and I agree. I'll open an issue to add it back in.

**Comment by charleskawczynski** [comment]
bors try

**Comment by charleskawczynski** [comment]
bors try

**Review by charleskawczynski — APPROVED** [review]

**Pull request merged**